File: leetcode_javascript/32.LongestValidParentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses2 = function(s) {
    let openCount = 0, total = 0, maxLen = 0;

    const getMaxLen = (isOpenChar) => {
            if (isOpenChar) {
                ++ openCount;
                ++ total;
            } else {
                -- openCount;
                ++ total;
            }

            if (openCount == 0) {
                maxLen = Math.max(maxLen, total);
            } else if (openCount < 0) {
                openCount = 0;
                total = 0;
            }
    }

    for (let i = 0; i < s.length; ++ i) {
        getMaxLen(s.charAt(i) === '(');
    }

    openCount = total = 0;
    for (let i = s.length - 1; i >= 0; -- i) {
        getMaxLen(s.charAt(i) === ')');
    }
    return maxLen;
};


/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    let openCount = 0, total = 0, maxLen = 0;

    for (let i = 0; i < s.length; ++ i) {
        getMaxLen(s.charAt(i) === '(');
    }

    openCount = total = 0;
    for (let i = s.length - 1; i >= 0; -- i) {
        getMaxLen(s.charAt(i) === ')');
    }
    return maxLen;
    
    function getMaxLen(isOpenChar) {
            if (isOpenChar) {
                ++ openCount;
                ++ total;
            } else {
                -- openCount;
                ++ total;
            }

            if (openCount == 0) {
                maxLen = Math.max(maxLen, total);
            } else if (openCount < 0) {
                openCount = 0;
                total = 0;
            }
    }
};


/**
32. Longest Valid Parentheses
Hard

Given a string containing just the characters '(' and ')', return the length of the longest valid (well-formed) parentheses.

 

Example 1:

Input: s = "(()"
Output: 2
Explanation: The longest valid parentheses substring is "()".

Example 2:

Input: s = ")()())"
Output: 4
Explanation: The longest valid parentheses substring is "()()".

Example 3:

Input: s = ""
Output: 0
 */

File: leetcode_python/32. Longest Valid Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        def getMaxLen(isOpen):
            nonlocal maxLen, openCount, totalCount
            openCount += 1 if isOpen else -1
            totalCount += 1
            
            if openCount == 0:
                maxLen = max(maxLen, totalCount)
            elif openCount < 0:
                totalCount = 0
                openCount = 0


        maxLen, openCount, totalCount = 0, 0, 0
        for ch in s:
            getMaxLen(ch == '(')
        
        openCount, totalCount = 0, 0
        for ch in s[::-1]:
            getMaxLen(ch == ')')

        return maxLen

        

/**
32. Longest Valid Parentheses
Solved
Hard
Topics
conpanies iconCompanies

Given a string containing just the characters '(' and ')', return the length of the longest valid (well-formed) parentheses .

 

Example 1:

Input: s = "(()"
Output: 2
Explanation: The longest valid parentheses substring is "()".

Example 2:

Input: s = ")()())"
Output: 4
Explanation: The longest valid parentheses substring is "()()".

Example 3:

Input: s = ""
Output: 0
 */
